packages/loopring_v3/contracts/aux/agents/FastWithdrawalAgent.sol

```diff
@@ -52,112 +58,148 @@ contract FastWithdrawalAgent is ReentrancyGuard
     using MathUint          for uint;
     using SignatureUtil     for bytes32;
 
-    struct FastWithdrawal
+    enum Status {
+        SUCCEEDED,
+        FRONTRUN,
+        EXPIRED,
+        TOO_LATE
+    }
+
+    event Processed(bytes32 hash, Status status);
+
+    struct FastWithdrawalApproval
     {
         address exchange;
         address from;                   // The owner of the account
         address to;                     // The address that will receive the tokens withdrawn
         address token;
         uint96  amount;
-        address feeToken;
-        uint96  fee;
-        uint32  nonce;
-        uint32  validUntil;
-
-        bytes   signature;
+        uint32  storageID;
+        int     validUntil; // seconds since epoch or block (if < 0)
+        address provider;
+        bytes   signature; // provider's signature
     }
 
-    // EIP712
-    bytes32 constant public FASTWITHDRAWAL_TYPEHASH = keccak256(
-        "FastWithdrawal(address exchange,address from,address to,address token,uint96 amount,address feeToken,uint96 fee,uint32 nonce,uint32 validUntil)"
+    bytes32 constant public FASTWITHDRAWAL_APPROVAL_TYPEHASH = keccak256(
+        "FastWithdrawalApproval(address exchange,address from,address to,address token,uint96 amount,uint32 storageID,int256 validUntil)"
     );
+
     bytes32 public DOMAIN_SEPARATOR;
 
+    mapping(bytes32 => bool) consumedApprovals;
```

Comment:
Doesn't seem necessary currently and makes things quite a bit more expensive.

My idea was to simply do this:
- Try/Catch `setWithdrawalRecipient`. In normal cases this will always succeed, only when the block already included the withdrawal does this fail. It also fails when trying to use the request more than once.
- Do the transfer if successful, don't do the transfer if not successful.

And that's it. Easy and efficient.

Reply:
We need to handle a replay issue. If a fast withdrawal approval is used and this tx is successful, after it is consumed by the protocol (which deletes the redirect), this same approval can be used again.

Reply:
@kongliangzhong is modifying this PR. redirect recipient will not be deleted any more, then we can remove consumedApprovals.

Reply:
> We need to handle a replay issue. If a fast withdrawal approval is used and this tx is successful, after it is consumed by the protocol (which deletes the redirect), this same approval can be used again.

That's why the protocol won't delete the redirect any more, so we can depend on it being there. 

> is modifying this PR. redirect recipient will not be deleted any more, then we can remove consumedApprovals.

Was already the case?